### PR TITLE
Ensure that ClickHouse will see ZooKeeper changes on config reloads

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Ensure that ClickHouse will see ZooKeeper changes on config reloads

Sometimes ClickHouse may not see ZooKeeper changes in case of multiple ZooKeeper replicas, i.e. if you created znode on one replica and ClickHouse connected to another it may not get the latest state. Issue `sync` before reloading the config to avoid this issues.

This should fix the `test_config_substitutions` test (see [1]).

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/71406/af822de0bf896dcb0a28f55889d12611bd38171a/integration_tests__asan__old_analyzer__[1_6].html

And also I made this test idempotent, otherwise it fails with `--count` [2].

  [2]: https://s3.amazonaws.com/clickhouse-test-reports/72593/422c42a52d675be4bd9ced067f56b4a3a80c195d/integration_tests_flaky_check__asan_.html

Refs: apache/zookeeper@8ee1732

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 422c42a52d675be4bd9ced067f56b4a3a80c195d with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72593/422c42a52d675be4bd9ced067f56b4a3a80c195d/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by Michicosun — COMMENTED** [review]
According to [zookeeper's doc](https://zookeeper.apache.org/doc/r3.8.4/zookeeperInternals.html#sc_consistency) there is no guarantee that read after sync will get up-to-date result and ticket is not closed: [ZOOKEEPER-1675](https://issues.apache.org/jira/browse/ZOOKEEPER-1675).

I think in the [test](https://s3.amazonaws.com/clickhouse-test-reports/71406/af822de0bf896dcb0a28f55889d12611bd38171a/integration_tests__asan__old_analyzer__[1_6].html) is more likely race between watch callback execution and config reloading thread.

btw ClickHouse Keeper runs sync through the raft log, so this guarantee is achieved there.

**Comment by azat** [comment]
>According to [zookeeper's doc](https://zookeeper.apache.org/doc/r3.8.4/zookeeperInternals.html#sc_consistency) there is no guarantee that read after sync will get up-to-date result and ticket is not closed: [ZOOKEEPER-1675](https://issues.apache.org/jira/browse/ZOOKEEPER-1675).

Thank you for the links! This is interesting, this is used for `ReplicatedMergeTree` and in other places, and I guess that those places expecting correct behavior.

>I think in the [test](https://s3.amazonaws.com/clickhouse-test-reports/71406/af822de0bf896dcb0a28f55889d12611bd38171a/integration_tests__asan__old_analyzer__%5B1_6%5D.html) is more likely race between watch callback execution and config reloading thread.

Hm, when I looked at the test logs I found that it did not find the znode

```
2024.11.27 20:53:47.654937 [ 9 ] {2c7313ba-d275-4ea3-a5b5-1a7e644df330} <Warning> ConfigProcessor: Could not get ZooKeeper node: /background_pool_size
```

So I don't see why this could be because of watches, the client (test.py) just created the `/background_pool_size` (note, previously it did not has `from_zk`, so it should not be cached), and then issued a `SYSTEM RELOAD CONFIG` so at this point server should see this znode, and

**Comment by alexey-milovidov** [comment]
What about making it unconditionally?

**Comment by azat** [comment]
>What about making it unconditionally?

The setting was added to avoid superior load to ZK in case of frequent reloads, but I'm not sure that this could be a problem.

OK, I will do a patch on top that remove the setting

**Comment by alexey-milovidov** [comment]
I think it's no problem at all.

**Comment by azat** [comment]
>[Integration tests (tsan) [6/6]](https://s3.amazonaws.com/clickhouse-test-reports/72593/195f359325161aa8b7e4cdaa673d227911458de8/integration_tests__tsan__[6_6].html)fail: 14, passed: 525

- https://github.com/ClickHouse/ClickHouse/issues/71049

>[AST fuzzer (debug)](https://s3.amazonaws.com/clickhouse-test-reports/72593/195f359325161aa8b7e4cdaa673d227911458de8/ast_fuzzer__debug_.html)Logical error: 'Inconsistent AST formatting: the query:

- https://github.com/ClickHouse/ClickHouse/issues/62975

File: tests/integration/helpers/cluster.py [hunk 1]
```diff
@@ -4542,6 +4543,20 @@ def _create_odbc_config_file(self):
                     if key != "DSN":
                         f.write(key + "=" + value + "\n")
 
+    @contextmanager
+    def with_replace_config(self, path, replacement):
+        """Create a copy of existing config (if exists) and revert on leaving the context"""
+        self.exec_in_container(
+            ["bash", "-c", f"test ! -f {path} || mv --no-clobber {path} {path}.bak"]
```
[the remark this answers is not in the document]
Comment (azat) [reply tests/integration/helpers/cluster.py:4550]: Make sense, done.

**Review by Michicosun — APPROVED** [review]
LGTM

**Comment by Michicosun** [comment]
failure in `Stress test (asan)` looks unrelated to the changes.

**Pull request merged**